api/src/infrastructure/docker.rs

```diff
@@ -986,55 +1169,108 @@ fn find_port(
             }),
         }
     } else {
-        Ok(match &container_details.network_settings.ports {
-            None => 80u16,
-            Some(ports) => {
-                let ports_regex = Regex::new(r#"^(?P<port>\d+).*"#).unwrap();
-                ports
-                    .keys()
-                    .filter_map(|port| ports_regex.captures(port))
-                    .map(|captures| String::from(captures.name("port").unwrap().as_str()))
-                    .filter_map(|port| port.parse::<u16>().ok())
-                    .min()
-                    .unwrap_or(80u16)
-            }
-        })
+        if let Some(network) = &container_details.network_settings {
+            Ok(network
+                .ports
+                .as_ref()
+                .map(|ports| {
+                    ports
+                        .iter()
+                        .filter_map(|(_, pb)| pb.as_ref())
+                        .flat_map(|bindings| bindings.iter())
+                        .filter_map(|binding| {
+                            binding
+                                .host_port
+                                .as_ref()
+                                .and_then(|hp| hp.parse::<u16>().ok())
+                        })
+                        .min()
+                })
+                .flatten()
+                .unwrap_or(80u16))
+        } else {
+            Ok(80u16)
+        }
     }
 }
 
-impl TryFrom<&ContainerDetails> for Service {
+impl TryFrom<&ContainerInspectResponse> for Service {
```

Comment:
In [my attempt](https://github.com/aixigo/PREvant/compare/master...replace-shiplift-with-bollard#diff-a9cc29c6f0841bd26edea1d6302aa4132b47529f46427c58af5f47865650fe1fR1011) I made this `TryFrom` more idiomatic and have less allocations. Please, check that out and merge our attempt. Also, it doesn't require ` let default_labels = HashMap::new();`

Reply:
I somehow missed this. I was going to type you a long message to understand how we can improve this.
I will understand and merge them.